Utils/src/L1NonPrefiringProbProducer.cc

```diff
@@ -0,0 +1,348 @@
+      double phi = muon.phi();
+      double eta = muon.eta();
+ 
+      // remove low quality muons that would not prefire the L1 trigger
+      if (pt < 5 || !muon.isLooseMuon())
+         continue;
+
+      const auto& nonprefiringprob_mu = getNonPrefiringRateMuon(eta, phi, pt);
+
+      dot(NonPrefiringProbsMuon, nonprefiringprob_mu);
+      dot(NonPrefiringProbs, nonprefiringprob_mu);
+   }
+
+   auto NonPrefiringProb = std::make_unique <double> ( NonPrefiringProbs[0]);
+   auto NonPrefiringProbUp = std::make_unique <double> ( NonPrefiringProbs[1]);
+   auto NonPrefiringProbDown = std::make_unique <double> ( NonPrefiringProbs[2]);
+   auto NonPrefiringProbECAL = std::make_unique <double> ( NonPrefiringProbsECAL[0]);
+   auto NonPrefiringProbECALUp = std::make_unique <double> ( NonPrefiringProbsECAL[1]);
+   auto NonPrefiringProbECALDown = std::make_unique <double> ( NonPrefiringProbsECAL[2]);
+   auto NonPrefiringProbMuon = std::make_unique <double> ( NonPrefiringProbsMuon[0]);
+   auto NonPrefiringProbMuonUp = std::make_unique <double> ( NonPrefiringProbsMuon[1]);
+   auto NonPrefiringProbMuonDown = std::make_unique <double> ( NonPrefiringProbsMuon[2]);
+
+   iEvent.put( std::move(NonPrefiringProb), "NonPrefiringProb" );
+   iEvent.put( std::move(NonPrefiringProbUp), "NonPrefiringProbUp" );
+   iEvent.put( std::move(NonPrefiringProbDown), "NonPrefiringProbDown" );
+   iEvent.put( std::move(NonPrefiringProbECAL), "NonPrefiringProbECAL" );
+   iEvent.put( std::move(NonPrefiringProbECALUp), "NonPrefiringProbECALUp" );
+   iEvent.put( std::move(NonPrefiringProbECALDown), "NonPrefiringProbECALDown" );
+   iEvent.put( std::move(NonPrefiringProbMuon), "NonPrefiringProbMuon" );
+   iEvent.put( std::move(NonPrefiringProbMuonUp), "NonPrefiringProbMuonUp" );
+   iEvent.put( std::move(NonPrefiringProbMuonDown), "NonPrefiringProbMuonDown" );
+}
+
+std::array<double,3> L1NonPrefiringProbProducer::getNonPrefiringRateECAL( double eta, double pt, TH2F * h_prefmap) const {
+   if(h_prefmap==0) return {{0.,0.,0.}};
+   
+   //Check pt is not above map overflow
+   int nbinsy = h_prefmap->GetNbinsY();
+   double maxy = h_prefmap->GetYaxis()->GetBinLowEdge(nbinsy+1);
+   if(pt>=maxy) pt = maxy-0.01;
+   int thebin = h_prefmap->FindBin(eta,pt);
+   
+   double prefrate = h_prefmap->GetBinContent(thebin);
+   double preferr = h_prefmap->GetBinError(thebin);
+   double prefrate_up = std::min(std::max(prefrate + preferr, (1. + prefiringRateSystUncECAL_)*prefrate),1.);
+   double prefrate_dn = std::max(std::min(prefrate - preferr, (1. - prefiringRateSystUncECAL_)*prefrate),0.);
+   
+   return {{1.-prefrate, 1.-prefrate_up, 1.-prefrate_dn}};
+}
+
+std::array<double,3> L1NonPrefiringProbProducer::getNonPrefiringRateMuon( double eta, double phi, double pt) const {
+
+   double abseta = abs(eta);
+   double prefrate = 0.0; double preferr = 0.0;
+   if ((dataeramuon_.find("2016") != std::string::npos) && (eta > 1.24 && eta < 1.6) && (phi > 2.44346 && phi < 2.79253)){
```

Comment:
string comparisons are expensive. since `dataeramuon_` is set in the constructor and does not change, just do this comparison in the constructor and keep a bool indicating that the hotspot function should be used.

Reply:
Done, in next commit

> 
> 
> > Would it be worth spending some time in the constructor to convert from TF1 to reco::FormulaEvaluator (https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/CommonTools/Utils/interface/FormulaEvaluator.h)? Since the TF1 would only be in the constructor, would that get around the need to use a stream over a global producer?
> 
> Yes, then a global producer could be used. As you know, this requires careful validation of all functional forms used to ensure that ROOT and FormulaEvaluator give consistent output. Given that, I probably would not prioritize it, but if someone wants to take it on, it would be nice to have (and maybe to contribute upstream to CMSSW).

For the moment, I just converted to an `edm::stream::EDProducer` along with addressing the other comments in e9088f6.